In [ ]:
import numpy as np
from scipy.stats import poisson
from scipy.stats import expon

In [ ]:
# You are working as a data engineer who has to resolve any bugs/
# failures of machine learning models in production.
# The time taken to debug is exponentially distributed with mean of 5 minutes

# Find the probability of debugging in 4 to 5 minutes?

In [ ]:
# 1 bug ----> 5 mins
#x bugs ------> 1 min
# x(rate)=1/5

#scale = 1/rate
#scale =1/(1/5)
#scale=5


In [ ]:
from scipy.stats import expon

# P[4<X<5]=CDF(5)-CDF(4)

expon.cdf(5,scale=5)-expon.cdf(4,scale=5)

np.float64(0.08144952294577923)

Q2 Find the probability of needing more than 6 minutes to debug.

In [ ]:
# P[X > 6]

1 - expon.cdf(x = 6, scale = 5)

np.float64(0.3011942119122022)

# Q3. Given that you have already spent 3 minutes, what is the probability of needing more than 9 minutes

In [ ]:
# P[T > 9 | T > 3]= P[T>9]/P[T>3] =P(T>6)

(1-expon.cdf(9,scale=5))/(1-expon.cdf(3,scale=5))

np.float64(0.30119421191220214)

The memoryless property essentially means that the time you've already spent on an event doesn't affect the future waiting time.
In the context of the exponential distribution, it means that the probability of needing more time in the future is the same, regardless of how much time has already passed.

# Practice Questions

A transportation analyst models flight delays (in minutes) at a major airport. She collects the following data:

- 70% of flights are on time (0 minutes delay)  
- 15% are delayed by exactly 30 minutes  
- 15% are delayed by exactly 60 minutes  

She reports this using a CDF plot to her non-technical team.

Later, another analyst claims:  
> "We can also derive a PDF and compute the probability that a flight is delayed by more than 30 minutes."

---

### Question:

What is the correct interpretation of this situation?

#### Options:

A. This is a **PDF** problem since the time is continuous  
B. The **CDF** allows computation of  
        $P(X > 30) = 1 - \text{CDF}(30) = 0.15$  
C. This is a **PMF**, so using CDF doesn’t make sense  
D. The correct probability for  
        $P(X > 30) = 0.3$


In [ ]:
#P(X>30) =1-cdf
#P(X<=30) =P(0)+P(30)
0.7+0.15

0.85

In [ ]:
1-0.85

0.15000000000000002

Question 2:

An insurance company receives 20 claims daily. The probability a claim is fraudulent is 0.15.

Find the probability that:
- A. At least 5 claims are fraudulent
- B. Exactly 3 claims are fraudulent
- C. Expected fraudulent claims and its variance

Options:

- A. A = 0.38, B = 0.25, Exp = 3, Var = 2.55
- B. A = 0.17, B = 0.24, Exp = 3, Var = 2.55
- C. A = 0.46, B = 0.25, Exp = 2, Var = 1.7
- D. A = 0.28, B = 0.26, Exp = 3, Var = 2.1

In [ ]:
#P(X>=5)=1-P(X<=4)
from scipy.stats import binom
1-binom.cdf(k=4,n=20,p=0.15)



np.float64(0.1701531537313461)

In [ ]:
#P(X=3)

# b)  P[X = 3]
binom.pmf(k=3, n= 20, p = 0.15)



np.float64(0.24282889614926753)

In [ ]:
20*0.15

3.0

In [ ]:
20 * 0.15 * (1 - 0.15)

2.55

Question 3:

Scenario:
A YouTuber earns revenue based on the number of ads watched. Each ad generates an amount between ₹0.50 and ₹3 randomly.
The revenue per ad follows a Uniform Distribution:
X
∼
U
(
0.5
,
3
)
X∼U(0.5,3)
The YouTuber uploads 5 videos, and each video gets an average of 1,000 ad views.

- Question 1: What is the expected revenue per ad?
- Question 2: What is the expected total revenue from all 5 videos?

Options:

A. Revenue per ad = ₹1.00, Total revenue = ₹5,000

B. Revenue per ad = ₹1.75, Total revenue = ₹8,750

C. Revenue per ad = ₹2.50, Total revenue = ₹12,500

D. Revenue per ad = ₹3.00, Total revenue = ₹15,000

In [ ]:
e= (0.5+3)/2

In [ ]:
e

1.75

In [ ]:
#num_videos*views_per_video*expected_revenue_per_ad
5*1000*1.75

8750.0

Scenario: A user plays a Google Pay cashback scratch card where they win ₹10, ₹20, ₹50, ₹100, or ₹500, each with an equal chance.

Question 1: What is the probability of winning ₹50? (Use PMF)

Question 2: What is the expected cashback a user will receive? (Use Expectation)


Options:

A. Probability = 0.25, Expected cashback = ₹136

B. Probability = 0.20, Expected cashback = ₹136

C. Probability = 0.10, Expected cashback = ₹100

D. Probability = 0.20, Expected cashback = ₹50

In [ ]:
#P(X=50) = 1/5

In [ ]:
np.mean([10,20,50,100,500])

np.float64(136.0)

Scenario: An Uber driver tracks ride durations in a city. Data shows that rides follow a normal distribution with:

Mean ride time (μ) = 15 minutes
Standard deviation (σ) = 3 minutes
Question 1: What is the probability that a randomly selected ride takes more than 20 minutes?

Question 2: What is the expected ride duration, and what is the variance?

Question 3: What is the probability that a ride takes between 12 and 18 minutes?


Options:

A. P(X>20)=0.05, Expected = 15 min, Variance = 9, P(12<X<18)=0.68

B. P(X>20)=0.10, Expected = 18 min, Variance = 3, P(12<X<18)=0.50

C. P(X>20)=0.03, Expected = 15 min, Variance = 6, P(12<X<18)=0.70

D. P(X>20)=0.15, Expected = 12 min, Variance = 9, P(12<X<18)=0.60

In [ ]:
#P(x>20)

from scipy.stats import norm

1-norm.cdf(20,15,3)

np.float64(0.047790352272814696)

In [ ]:
#sd= sqrt(var)
var=3*3

In [ ]:
var

9

In [ ]:
p_between_12_18 = norm.cdf(18, 15,3) - norm.cdf(12, 15, 3)

In [ ]:
p_between_12_18

np.float64(0.6826894921370859)

# Suppose you are playing a game where success rate of winning a prize is 0.7.
# (a) What is the probability of winning a prize on 4th attempt?
# (b) What is the probability that you don't win in 2 attempts?
# (c) What is the expected number of trials to get the first success?

In [ ]:
from scipy.stats import geom

In [ ]:
geom.pmf(k=4,p=0.7)

np.float64(0.018900000000000007)

In [ ]:
#P[X>2]=1-cdf(2)
1-geom.cdf(k=2,p=0.7)

np.float64(0.09000000000000008)

In [ ]:
e=1/(0.7)

In [ ]:
e

1.4285714285714286

Question 11:

Scenario: A WhatsApp user receives an average of 5 messages per minute. The number of messages follows a Poisson distribution with λ=5.

Question 1: What is the probability that the user receives exactly 7 messages in a minute?

Question 2: What is the probability that the user receives at least 3 messages in a minute?


A. P(X=7) = 0.1044, P(X≥3) = 0.8753

B. P(X=7) = 0.1200, P(X≥3) = 0.8600

C. P(X=7) = 0.0900, P(X≥3) = 0.8900

D. P(X=7) = 0.1100, P(X≥3) = 0.8800


In [ ]:
poisson.pmf(k=7,mu= 5)

np.float64(0.10444486295705395)

In [ ]:
#P[X>=3]=1-poisson.cdf(2)

1-poisson.cdf(k=2,mu=5)

np.float64(0.8753479805169189)

A food delivery app records the time taken for deliveries in minutes:

50% of the orders are delivered in 25 minutes.

30% take 35 minutes.

20% take 45 minutes.

Question 1: What is the expected delivery time?

Question 2: If a customer places 3 independent orders, what is the expected total delivery time?

Options:

A. Expected time = 30 minutes, Total time = 90 minutes

B. Expected time = 32 minutes, Total time = 96 minutes

C. Expected time = 35 minutes, Total time = 105 minutes

D. Expected time = 40 minutes, Total time = 120 minutes

In [ ]:
# Given data
times = np.array([25, 35, 45])  # Delivery times in minutes
probabilities = np.array([0.5, 0.3, 0.2])  # Probabilities

# 1️⃣ Expected delivery time (E(X) = ∑x_i * P(x_i))
expected_delivery_time = np.sum(times * probabilities)

# 2️⃣ Expected delivery time for 3 orders (Linearity: E(3X) = 3 * E(X))
expected_total_time_3_orders = 3 * expected_delivery_time

print(f"Expected delivery time for one order: {expected_delivery_time:.2f} minutes")
print(f"Expected total delivery time for 3 orders: {expected_total_time_3_orders:.2f} minutes")


Expected delivery time for one order: 32.00 minutes
Expected total delivery time for 3 orders: 96.00 minutes


## Homework (10 questions)

Q1. The weight of packaged rice bags is normally distributed with a mean of 1 kg and standard deviation of 0.05 kg.
A retailer rejects bags that weigh less than 0.92 kg or more than 1.08 kg. What proportion of bags will be rejected?


Q2. IQ scores are normally distributed with mean 100 and standard deviation 15.
A university only accepts applicants in the top 5% based on IQ.What is the minimum IQ score required for admission?


Q3. A company claims their batteries last an average of 400 hours with a standard deviation of 30 hours.
In a random sample of 100 batteries, what is the probability that the average life is less than 395 hours? (Use CLT)


Q4. In a quality control process, each item has a 90% chance of passing inspection.
Out of 15 items, the batch is accepted if at least 13 pass.What is the probability that a batch is accepted?



Q5. A student guesses answers to a 10-question true/false quiz.What is the probability the student scores at least 8 correct answers by guessing?


Q6. A biased die has a 1/4 probability of rolling a 6. The die is rolled 12 times.What is the expected number of 6s, and what is the probability of getting exactly 3?


Q7. A hospital emergency room receives an average of 6 patients per hour.What is the probability that in a given 30-minute interval, at most 2 patients arrive?


Q8. A rare event occurs with an average rate of 0.5 times per day. What is the probability that it occurs on 3 or more days in a 5-day work week?

Q9. A basketball player has a 60% chance of making a free throw.What is the probability that she misses her first 2 shots but makes the 3rd one?

Q10. An IT support ticket has a 30% chance of being resolved on the first attempt. What is the probability it takes more than 4 attempts to resolve an issue?
